arch/rv32i/src/syscall.rs

```diff
@@ -0,0 +1,397 @@
+//! Kernel-userland system call interface for RISC-V architecture.
+
+use core::fmt::Write;
+
+use kernel;
+
+/// This holds all of the state that the kernel must keep for the process when
+/// the process is not executing.
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+pub struct RiscvimacStoredState {
+    /// Store all of the app registers.
+    regs: [usize; 31],
+
+    /// This holds the PC value of the app when the exception/syscall/interrupt
+    /// occurred. We also use this to set the PC that the app should start
+    /// executing at when it is resumed/started.
+    pc: usize,
+
+    /// We need to store the mcause CSR between when the trap occurs and after
+    /// we exit the trap handler and resume the context switching code.
+    mcause: usize,
+}
+
+/// Implementation of the `UserspaceKernelBoundary` for the RISC-V architecture.
+pub struct SysCall(());
+
+impl SysCall {
+    pub const fn new() -> SysCall {
```

Comment:
Is it, or is it not safe to have multiple instances of `SysCall`? If it is, than this is fine, but the definition of `SysCall` doesn't need to have any extra parens or unit types, just a plain `pub struct SysCall;` is good enough.

If it _does_, then this _has_ to be marked `unsafe`.

Reply:
This is a discrepancy we have in the cortex-m variants as well (where this was likely copied from). Let's decide here and fix in the cortex-m in a different PR.

Reply:
It certainly doesn't make a lot of sense to have multiple `SysCall` objects, but I'm not sure if it is strictly safe or not. Since everything is handled on the stack it might be safe.

Reply:
We typically think of it as potentially unsafe when we have might end up with two instances that use the same set of hardware registers, so that they are not _actually_ isolated instances...

I think in this particular case, I think `switch_to_process` is what should motivate us to ensure `SysCall` is a singleton and keep `new` unsafe: It's operating on a shared hardware variables, such as `mscratch`, `mepc` etc...

Reply:
I agree.